### ISSUE TITLE
feat: checkpoint mechanism for TracedTensor (rematerialization)

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#690]
## Summary

Add a checkpoint mechanism to `TracedTensor` that enables memory-efficient backward passes through iterative computations (TN sweeps, fixed-point iterations, etc.). This is the tenferro-rs equivalent of `jax.checkpoint` / `torch.utils.checkpoint`.

## Motivation

When differentiating through L-step iterative algorithms (e.g., TN sweep with L sites), backward needs primal intermediate values at every step. Without checkpoint:
- **Store all**: O(L) memory
- **Recompute all**: O(L²) compute (each step's eval recomputes from scratch — see #654 pattern)

With √L checkpoints:
- **O(√L) memory, O(3L) compute** (1.5x overhead)

Current `TracedTensor::eval()` does not break graph connectivity (#654 verified), but mid-loop eval has O(N²) cost because intermediate `.data` is not cached and the resolver always traces back to original inputs.

### Core API

```rust
impl TracedTensor {
    /// Mark this tensor as a checkpoint boundary.
    ///
    /// 1. Evaluates the tensor (computes concrete value)
    /// 2. Saves the concrete value
    /// 3. Promotes to a new leaf input (cuts the forward graph here)
    /// 4. Preserves the old graph in `extra_roots` for backward recomputation
    ///
    /// After checkpoint:
    /// - Forward: subsequent ops start from this leaf (no recomputation of prior ops)
    /// - Backward: recomputes from this checkpoint to the previous checkpoint
    pub fn checkpoint(&mut self, engine: &mut Engine<impl TensorBackend>) -> Result<()>;
}
```

### Implementation

```rust
pub fn checkpoint(&mut self, engine: &mut Engine<impl TensorBackend>) -> Result<()> {
    // 1. Evaluate to get concrete data
    self.eval(engine)?;
    
    // 2. Promote to leaf: create a new Fragment with this tensor as input
    let data = self.data.clone().expect("eval should populate data");
    let new_key = TensorInputKey::materialized(self.id);
    
    let mut builder = FragmentBuilder::new();
    let val = builder.add_input(new_key.clone());
    builder.set_outputs(vec![val]);
    
    // 3. Preserve old graph for backward
    let old_fragment = std::mem::replace(&mut self.fragment, Arc::new(builder.build()));
    self.extra_roots.push(old_fragment);
    
    // 4. Update inputs_map with concrete data
    let mut new_inputs = (*self.inputs_map).clone();
    new_inputs.insert(new_key, data);
    self.inputs_map = Arc::new(new_inputs);
    self.val = val;
    
    Ok(())
}
```

### Key behaviors

1. **Forward graph cut**: After `checkpoint()`, the TracedTensor is a leaf. Subsequent ops build a new graph from here. `eval()` on downstream tensors only computes from the checkpoint, not from the original inputs.

2. **Backward graph intact**: `extra_roots` holds the old fragment. When `grad()` is called, `differentiate()` can still trace back through the old graph to compute gradients.

3. **Per-step eval becomes O(K)**: With checkpoint at step k, step k+1's graph starts from the checkpoint (leaf). Evaluating step k+1's outputs only computes K ops (one step), not k×K ops (all prior steps).

### AD through checkpoints

When `loss.grad(&input)` is called after a checkpointed loop:
1. `differentiate()` walks the backward graph through `extra_roots`
2. Between checkpoints, the forward ops are included in the backward graph for recomputation
3. The merged `eval_all([loss, grad])` execution:
   - Uses checkpoint concrete values as inputs to each segment
   - Recomputes forward within each segment
   - Computes gradients using the recomputed primals

### Verification needed

- [ ] `checkpoint()` preserves graph connectivity for `grad()`
- [ ] After checkpoint, subsequent `eval()` is O(K) not O(k×K)  
- [ ] `eval_all([primal, grad])` produces correct gradients through checkpointed loop
- [ ] Finite-difference validation of gradients through checkpointed fixed-point iteration
- [ ] √N checkpoint spacing: memory is O(√N), compute is O(3N)
- [ ] Graph size: fragment count after checkpoint is bounded (not growing unboundedly)

### Open questions

- How does `extra_roots` interact with `resolve()` for the backward graph? Need to verify that the old fragments are discoverable during `differentiate()`.
- `TensorInputKey::materialized(id)` — does this key type exist? May need a new variant.
- Interaction with `eval_all`: when multiple tensors are checkpointed in the same step, ensure they share the same eval (not N separate evals).

## Related

- #654 — iterative AD with eval() mid-loop (validates graph connectivity)
- `docs/design/symbolic-shape-and-tenferro-migration.md` — Phase 4 design

**Comment by shinaoka** [comment]
## Design update: zero-copy checkpoint via Arc<Tensor>

### Problem

The current `TracedTensor` stores:
```rust
pub data: Option<Tensor>,                         // concrete value after eval
pub(crate) inputs_map: Arc<HashMap<TensorInputKey, Tensor>>,  // leaf inputs
```

`checkpoint()` needs to register the eval'd `Tensor` as a new leaf input in `inputs_map`. With the current types, this requires cloning `Tensor` (= `Vec<T>` full copy = O(elements) memory + time).

### Solution: Arc<Tensor> throughout

```rust
// Before
pub data: Option<Tensor>,
pub(crate) inputs_map: Arc<HashMap<TensorInputKey, Tensor>>,

// After
pub data: Option<Arc<Tensor>>,
pub(crate) inputs_map: Arc<HashMap<TensorInputKey, Arc<Tensor>>>,
```

Checkpoint becomes zero-copy:
```rust
fn checkpoint(&mut self, engine: &mut Engine) -> Result<()> {
    self.eval(engine)?;
    let data: Arc<Tensor> = self.data.clone().unwrap();  // Arc clone = refcount increment only

    let new_key = TensorInputKey::materialized(self.id);
    let mut new_inputs = HashMap::new();
    new_inputs.insert(new_key.clone(), data);  // zero copy

    self.extra_roots.push(self.fragment.clone());  // preserve old graph for AD

    let mut builder = FragmentBuilder::new();
    let val = builder.add_input(new_key);
    builder.set_outputs(vec![val]);
    self.fragment = Arc::new(builder.build());
    self.val = val;
    self.inputs_map = Arc::new(new_inputs);
    Ok(())
}
```

Cost: **Arc refcount increment only. No memory copy.**

### Public API impact

`eval()` returns `&Tensor` — unchanged via `Arc::as_ref()`:
```rust
pub fn eval(&mut self, engine: &mut Engine) -> Result<&Tensor> {
    // ...
    self.data = Some(Arc::new(result));
    Ok(self.data.as_ref().unwrap().as_ref())
}
```

### EagerTensor as sugar

With checkpoint implemented, EagerTensor is a thin wrapper:
```rust
struct EagerTensor {
    traced: TracedTensor,
}

impl EagerTensor {
    fn materialize(&mut self, engine: &mut Engine) -> Result<()> {
        self.traced.eval(engine)?;
        self.traced.checkpoint(engine)?;  // zero-copy via Arc
        Ok(())
    }
}
```

Per-step materialize: each step's graph is compiled (cache hit after step 1), executed, and checkpointed. O(N) total, zero unnecessary memory copies.

### Implementation prerequisite

`Arc<Tensor>` change touches:
- `TracedTensor::data` field type
- `TracedTensor::eval()` — wrap result in Arc
- `eval_all()` — same
- `inputs_map` value type
- Input resolution in eval (Arc deref)
- `try_vjp` / `try_jvp` where `data` is accessed

This should be done as a separate PR before checkpoint implementation.

**Comment by shinaoka** [comment]
Implemented by #692 (merged). `TracedTensor::checkpoint()` with `Arc<CheckpointNode>` persistent linked list, full AD support.
